geo: order linear rings correctly when converting to S2

otan opened the pull request:
We falsely assumed all WKTs we were given would be CCW. Some could be
CW, as is the case with ESRI shape files. Since S2 requires CCW, we now
check the linear ring is CCW with math.

Note that the geoindex tests still have an invalid result as the WKT for
New York State is invalid.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47573)
<!-- Reviewable:end -->

otan commented:
the WKT for NY state is invalid:

```

otan=# select ST_IsValid('POLYGON((-79.7624 42.5142,-79.0672 42.7783,-78.9313 42.8508,-78.9024 42.9061,-78.9313 42.9554,-78.9656 42.9584,-79.0219 42.9886,-79.0027 43.0568,-79.0727 43.0769,-79.0713 43.1220,-79.0302 43.1441,-79.0576 43.1801,-79.0604 43.2482,-79.0837 43.2812,-79.2004 43.4509,-78.6909 43.6311,-76.7958 43.6321,-76.4978 43.9987,-76.4388 44.0965,-76.3536 44.1349,-76.3124 44.1989,-76.2437 44.2049,-76.1655 44.2413,-76.1353 44.2973,-76.0474 44.3327,-75.9856 44.3553,-75.9196 44.3749,-75.8730 44.3994,-75.8221 44.4308,-75.8098 44.4740,-75.7288 44.5425,-75.5585 44.6647,-75.4088 44.7672,-75.3442 44.8101,-75.3058 44.8383,-75.2399 44.8676,-75.1204 44.9211,-74.9995 44.9609,-74.9899 44.9803,-74.9103 44.9852,-74.8856 45.0017,-74.8306 45.0153,-74.7633 45.0046,-74.7070 45.0027,-74.5642 45.0007,-74.1467 44.9920,-73.7306 45.0037,-73.4203 45.0085,-73.3430 45.0109,-73.3547 44.9874,-73.3379 44.9648,-73.3396 44.9160,-73.3739 44.8354,-73.3324 44.8013,-73.3667 44.7419,-73.3873 44.6139,-73.3736 44.5787,-73.3049 44.4916,-73.2953 44.4289,-73.3365 44.3513,-73.3118 44.2757,-73.3818 44.1980,-73.4079 44.1142,-73.4367 44.0511,-73.4065 44.0165,-73.4079 43.9375,-73.3749 43.8771,-73.3914 43.8167,-73.3557 43.7790,-73.4244 43.6460,-73.4340 43.5893,-73.3969 43.5655,-73.3818 43.6112,-73.3049 43.6271,-73.3063 43.5764,-73.2582 43.5675,-73.2445 43.5227,-73.2582 43.2582,-73.2733 42.9715,-73.2898 42.8004,-73.2664 42.7460,-73.3708 42.4630,-73.5095 42.0840,-73.4903 42.0218,-73.4999 41.8808,-73.5535 41.2953,-73.4834 41.2128,-73.7275 41.1011,-73.6644 41.0237,-73.6578 40.9851,-73.6132 40.9509,-72.4823 41.1869,-72.0950 41.2551,-71.9714 41.3005,-71.9193 41.3108,-71.7915 41.1838,-71.7929 41.1249,-71.7517 41.0462,-72.9465 40.6306,-73.4628 40.5368,-73.8885 40.4887,-73.9490 40.5232,-74.2271 40.4772,-74.2532 40.4861,-74.1866 40.6468,-74.0547 40.6556,-74.0156 40.7618,-73.9421 40.8699,-73.8934 40.9980,-73.9854 41.0343,-74.6274 41.3268,-74.7084 41.3583,-74.7101 41.3811,-74.8265 41.4386,-74.9913 41.5075,-75.0668 41.6000,-75.0366 41.6719,-75.0545 41.7672,-75.1945 41.8808,-75.3552 42.0013,-75.4266 42.0003,-77.0306 42.0013,-79.7250 41.9993,-79.7621 42.0003,-79.7621 42.1827,-79.7621 42.5146,-79.7624 42.5142))');
NOTICE:  Self-intersection at or near point -79.762100000000004 42.514313967203684
 st_isvalid
------------
 f
(1 row)
```

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geo.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5HrfHX14chHMEsIeHu:-M5HrfHX14chHMEsIeHv:b-1jb7ez) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a873285db2a2734341cc79f553cab00f8c80dc4/pkg/geo/geo.go#L191)):*
> ```Go
> // If a < b, then we have a "left turn". Solving that, we need to solve
> // (y2 - y1)/(x2 - x1) < (y3 - y2)/(x3 - x2) - or rearranged -
> // (y2 - y1) * (x3 - x2) - (y3 - y2) * (x2 - x1) < 0
> ```

I don't understand a few things here:
- isn't is possible for a transition from positive to negative slope also be a "left turn". So say p1=(0,0), p2=(5,5,), p3=(10,0). 
- continuing this example, p4=(15,-10). The slope changes from -1 to -1/2 which is an increase, but I think this also represents a "left turn".
- IIUC, this "left turn" idea is capturing only a convex ring. We are allowed to have non-convex rings which will have a mix of "left turns" and "right turns". It seems to me that we can't really fix the data in our code -- we need to assume that what is provided is what the user intended.


<!-- Sent from Reviewable.io -->

otan commented:
Oops this is the wrong version o_o thought I submitted an update

otan committed:
geo: order linear rings correctly when converting to S2

We falsely assumed all WKTs we were given would be CCW. Some could be
CW, as is the case with ESRI shape files. Since S2 requires CCW, we now
check the linear ring is CCW with math.

Note that the geoindex tests still have an invalid result as the WKT for
New York State is invalid.

Release note: None

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5OGTsd5eCQ-rmfS87d:busl4l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
> ```Go
> 
> // isLinearRingCCW returns whether a given linear ring is counter clock wise.
> // See 2.07 of http://www.faqs.org/faqs/graphics/algorithms-faq/.
> ```

I believe this for 2D planar, but I am not sure about 2D sphere -- it seems to me that the difference between a polygon that represents the USA and the one that represents "Earth except for USA" is only the orientation of the points and both are valid polygons.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geo.go, line 191 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5HrfHX14chHMEsIeHu:-M5PBw_DAWQsEuWY2r_F:bqa14tw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5a873285db2a2734341cc79f553cab00f8c80dc4/pkg/geo/geo.go#L191)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't understand a few things here:
- isn't is possible for a transition from positive to negative slope also be a "left turn". So say p1=(0,0), p2=(5,5,), p3=(10,0). 
- continuing this example, p4=(15,-10). The slope changes from -1 to -1/2 which is an increase, but I think this also represents a "left turn".
- IIUC, this "left turn" idea is capturing only a convex ring. We are allowed to have non-convex rings which will have a mix of "left turns" and "right turns". It seems to me that we can't really fix the data in our code -- we need to assume that what is provided is what the user intended.
</blockquote></details>

I think this is fixed with the new change. Marking as resolved.
___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5PC04F2ga5HhhWGUWo:bpagl5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I believe this for 2D planar, but I am not sure about 2D sphere -- it seems to me that the difference between a polygon that represents the USA and the one that represents "Earth except for USA" is only the orientation of the points and both are valid polygons.
</blockquote></details>

It should be roughly the same for sphere. I'm not sure "earth except for USA" is a valid polygon in PostGIS terms (you can't have a global covering https://postgis.net/docs/using_postgis_dbmanagement.html#OGC_Validity) - all polygons must be "Simple".

This change makes data work with the tutorial, so I'm inclined to keep it around.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5PQV7s4CKxTDKuZAXt:b-7xhhim) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

It should be roughly the same for sphere. I'm not sure "earth except for USA" is a valid polygon in PostGIS terms (you can't have a global covering https://postgis.net/docs/using_postgis_dbmanagement.html#OGC_Validity) - all polygons must be "Simple".

This change makes data work with the tutorial, so I'm inclined to keep it around.
</blockquote></details>

I'm not concerned about the tutorial if the data is incorrect.

And I'm not clear on whether "simple" applies to this case. Could you elaborate on what aspect of "simple" is violated?

S2 definitely has the behavior that I was asking about, and my impression was that this was standard practice in sphere geometry.
From the S2 documentation:
`An S2Loop represents a simple spherical polygon. It consists of a single chain of vertices where the first vertex is implicitly connected to the last. All loops are defined to have a CCW orientation, i.e. the interior of the loop is on the left side of the edges. This implies that a clockwise loop enclosing a small area is interpreted to be a CCW loop enclosing a very large area.
`

There is a normalization function that users who want to "fix" their data can use

`
When the orientation of the input loops is unknown, ... calling S2Loop::Normalize() on each loop (which inverts the loop if necessary so that it encloses at most half the sphere). 
`

Can you confirm that PostGIS does not allow this?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5PbO8vDPYc7JRtadqD:bwpg3n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm not concerned about the tutorial if the data is incorrect.

And I'm not clear on whether "simple" applies to this case. Could you elaborate on what aspect of "simple" is violated?

S2 definitely has the behavior that I was asking about, and my impression was that this was standard practice in sphere geometry.
From the S2 documentation:
`An S2Loop represents a simple spherical polygon. It consists of a single chain of vertices where the first vertex is implicitly connected to the last. All loops are defined to have a CCW orientation, i.e. the interior of the loop is on the left side of the edges. This implies that a clockwise loop enclosing a small area is interpreted to be a CCW loop enclosing a very large area.
`

There is a normalization function that users who want to "fix" their data can use

`
When the orientation of the input loops is unknown, ... calling S2Loop::Normalize() on each loop (which inverts the loop if necessary so that it encloses at most half the sphere). 
`

Can you confirm that PostGIS does not allow this?
</blockquote></details>

Misread your question. I'm not totally sure you can represent the "whole world except USA" in PostGIS as it always considers the outer ring to be closed inside.

I just tried "Normalize" and it is giving me invalid coverings for everything.  The polygon depicted below is totally broken.

`drawgeography,The Rockaways,,"MULTIPOLYGON(((-73.7662110621911 40.6206568820531,-73.7659731160949 40.6149461757457,-73.7619280324605 40.612090822592,-73.7538378651916 40.6106631460151,-73.751220458134 40.6128046608804,-73.7462235901149 40.6123287686881,-73.7424164525766 40.6092354694382,-73.7378954767499 40.6028109248424,-73.7388472611345 40.5985278951118,-73.7369436923653 40.5954345958619,-73.7357606184878 40.593512100811,-73.7455097518265 40.5930551349005,-73.7516963503263 40.5890100512661,-73.7726356067871 40.5897238895545,-73.8064239524396 40.5828234527663,-73.8461609504958 40.5721158784399,-73.8951778463016 40.5561734899983,-73.9035059596666 40.5566493821906,-73.9420532272421 40.542610562518,-73.9396737662807 40.5552217056137,-73.9329531744502 40.5577569082933,-73.9303938685311 40.5621221424019,-73.9168309410508 40.5635498189787,-73.9093277021897 40.5663946488786,-73.9056474745319 40.5645016033633,-73.9011830440826 40.5662995651142,-73.8917112402415 40.5701141482311,-73.8809010805329 40.5704502557669,-73.8677581997444 40.5762036745353,-73.8497604278618 40.5840823672058,-73.839974351996 40.5830613988625,-73.8076136829203 40.5966243263427,-73.8052342219589 40.6011453021694,-73.7954784320169 40.6006694099771,-73.7874798715818 40.6061318414937,-73.7766806904215 40.6113769843035,-73.7734734771453 40.6142197415256,-73.7662110621911 40.6206568820531)))"`


We can try bench this for now and add something later that forces everything to be CCW. I'm a little weary of having tutorials not work out of the box.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5XSBuz4wfFPsp-pXZp:b-ivo8ge) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Misread your question. I'm not totally sure you can represent the "whole world except USA" in PostGIS as it always considers the outer ring to be closed inside.

I just tried "Normalize" and it is giving me invalid coverings for everything.  The polygon depicted below is totally broken.

`drawgeography,The Rockaways,,"MULTIPOLYGON(((-73.7662110621911 40.6206568820531,-73.7659731160949 40.6149461757457,-73.7619280324605 40.612090822592,-73.7538378651916 40.6106631460151,-73.751220458134 40.6128046608804,-73.7462235901149 40.6123287686881,-73.7424164525766 40.6092354694382,-73.7378954767499 40.6028109248424,-73.7388472611345 40.5985278951118,-73.7369436923653 40.5954345958619,-73.7357606184878 40.593512100811,-73.7455097518265 40.5930551349005,-73.7516963503263 40.5890100512661,-73.7726356067871 40.5897238895545,-73.8064239524396 40.5828234527663,-73.8461609504958 40.5721158784399,-73.8951778463016 40.5561734899983,-73.9035059596666 40.5566493821906,-73.9420532272421 40.542610562518,-73.9396737662807 40.5552217056137,-73.9329531744502 40.5577569082933,-73.9303938685311 40.5621221424019,-73.9168309410508 40.5635498189787,-73.9093277021897 40.5663946488786,-73.9056474745319 40.5645016033633,-73.9011830440826 40.5662995651142,-73.8917112402415 40.5701141482311,-73.8809010805329 40.5704502557669,-73.8677581997444 40.5762036745353,-73.8497604278618 40.5840823672058,-73.839974351996 40.5830613988625,-73.8076136829203 40.5966243263427,-73.8052342219589 40.6011453021694,-73.7954784320169 40.6006694099771,-73.7874798715818 40.6061318414937,-73.7766806904215 40.6113769843035,-73.7734734771453 40.6142197415256,-73.7662110621911 40.6206568820531)))"`


We can try bench this for now and add something later that forces everything to be CCW. I'm a little weary of having tutorials not work out of the box.
</blockquote></details>

is this up to date based on yesterday's discussion?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 186 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-M5OGTsd5eCQ-rmfS87c:-M5PhfWg7R2OXEtHSM0H:b-by94mx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c542f329e7cf79bcd66c6c250d7bc1eb78578ea2/pkg/geo/geo.go#L186)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this up to date based on yesterday's discussion?
</blockquote></details>

Yes. The `Normalize()` from S2 doesn't work, and this is the only code I can make that seems to make this work.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47573#-:-M5h-GT9BOsxHCcVZNea:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
